### PR TITLE
build: build the section_magic libraries for ELFish targets

### PR DESCRIPTION
**Pull request by compnerd**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

Build the section_magic libraries for all the ELFish SDKs that have been
configured.  This allows us to build for SDKs which dont match the object format
at the same time (e.g. Linux and Windows).

### CODE REVIEW

**Comment by compnerd** [comment]
CC @modocache

File: stdlib/public/runtime/CMakeLists.txt [hunk 1]
```diff
@@ -82,16 +82,24 @@ add_swift_library(swiftRuntime OBJECT_LIBRARY TARGET_LIBRARY
...
+set(ELFISH_SDKS)
+foreach(sdk ${SWIFT_CONFIGURED_SDKS})
+  is_elfish_sdk("${sdk}" IS_ELFISH)
+  if(IS_ELFISH)
+    list(APPEND ELFISH_SDKS ${sdk})
+  endif()
+endforeach()
+
+add_swift_library(section_magic_begin STATIC IS_STDLIB IS_STDLIB_CORE
+                  swift_sections.S
```
Comment (gribozavr) [inline stdlib/public/runtime/CMakeLists.txt]: Please use 4 spaces on continuation lines.

**Comment by gribozavr** [comment]
I think there might be an issue with this code in the same file below, `add_custom_command_target(section_magic_o` -- there can be only one target with a given name.

**Comment by compnerd** [comment]
I think that you are right about that, but that seems to be a bug that has been there for a while?  Am I just going crazy or am I misunderstanding something?  (This is about trying to understand whats going on, Im happy to try to help fix the issue).

**Comment by gribozavr** [comment]
@compnerd No, I think we had this bug for a while -- it is just that with your change we are increasing the possibility of creating multiple `section_magic_o` targets.  Would be totally fair to fix it in a separate PR.

File: stdlib/public/runtime/CMakeLists.txt [hunk 2]
```diff
@@ -82,26 +82,37 @@ add_swift_library(swiftRuntime OBJECT_LIBRARY TARGET_LIBRARY
...
     C_COMPILE_FLAGS ${swift_runtime_compile_flags} "-DSWIFT_END"
+    TARGET_SDKS "${ELFISH_SDKS}"
     INSTALL_IN_COMPONENT never_install)
-endif()
 
 set(object_target_list)
 foreach(sdk ${SWIFT_CONFIGURED_SDKS})
-  if("${sdk}" STREQUAL "LINUX" OR "${sdk}" STREQUAL "FREEBSD" OR "${sdk}" STREQUAL "ANDROID")
+  is_elfish_sdk("${sdk}" IS_ELFISH)
+  if(IS_ELFISH)
```
Comment (gribozavr) [inline stdlib/public/runtime/CMakeLists.txt]: A random suggestion, but adding `${SWIFT_SDK_${sdk}_OBJECT_FORMAT}` would help to write these conditions more compactly:

```cmake
if("${SWIFT_SDK_${sdk}_OBJECT_FORMAT}" STREQUAL "ELF")
```

**Comment by gribozavr** [comment]
@compnerd I see you fixed it.  Great, thank you!

@swift-ci Please test and merge

File: stdlib/public/runtime/CMakeLists.txt [hunk 3]
```diff
@@ -82,26 +82,37 @@ add_swift_library(swiftRuntime OBJECT_LIBRARY TARGET_LIBRARY
...
       set(arch_subdir "${SWIFT_SDK_${sdk}_LIB_SUBDIR}/${arch}")
+      set(arch_suffix "${SWIFT_SDK_${sdk}_LIB_SUBDIR}-${arch}")
+
+      set(section_magic_begin_name "section_magic_begin-${arch_suffix}")
+      set(section_magic_end_name "section_magic_end-${arch_suffix}")
 
-      set(section_magic_begin_name "section_magic_begin-${SWIFT_SDK_${sdk}_LIB_SUBDIR}-${arch}")
-      set(section_magic_end_name "section_magic_end-${SWIFT_SDK_${sdk}_LIB_SUBDIR}-${arch}")
-      add_custom_command_target(section_magic_o
+      add_custom_command_target(section_magic_${arch_suffix}_objects
```
Comment (rintaro) [inline stdlib/public/runtime/CMakeLists.txt]: `section_magic_o` was just a variable name that hold the generated target name *only while* single loop iteration.
I don't think we need to use unique name for this.
Consider this pseudo code:

```swift
var object_target_list = []
for arch in arch_list {
    let section_magic_o = add_custom_command_target( ... )
    object_target_list.append(section_magic_o)
}
```

**Pull request merged**